src/NHibernate/Engine/TableGroupJoinHelper.cs

```diff
@@ -68,14 +68,15 @@ private static bool NeedsTableGroupJoin(IReadOnlyList<IJoin> joins, SqlString[]
 
 			foreach (var join in joins)
 			{
-				var entityPersister = GetEntityPersister(join.Joinable);
+				var entityPersister = GetEntityPersister(join.Joinable, out var collection);
 				if (entityPersister?.HasSubclassJoins(includeSubclasses && isSubclassIncluded(join.Alias)) != true)
 					continue;
 
 				if (hasWithClause)
 					return true;
 
-				if (entityPersister.ColumnsDependOnSubclassJoins(join.RHSColumns))
+				if (collection?.IsManyToMany != true // many-to-many keys are stored in separate table
+				    && entityPersister.ColumnsDependOnSubclassJoins(join.RHSColumns))
 					return true;
```

Comment:
Please add braces here as it is hard to read now.

Reply:
Where exactly? Is it still valid?

Reply:
Yes, still valid. Around `return true;`
```suggestion
				{
					return true;
				}
```